packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,67 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL, SLOT_COMPONENT_TYPEOF_SYMBOL } from './constants';
+import * as React from 'react';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  componentType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+  overrides?: Partial<Props>;
+};
+
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | undefined,
+  options: SlotOptions<Props> & { required: true },
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, overrides, componentType } = options;
+
+  if (shorthand === null || (shorthand === undefined && !required)) {
+    return undefined;
+  }
+  const metadata: SlotComponentMetadata<Props> = { componentType };
+
+  const slotComponent = {
+    ...defaultProps,
+    $$typeof: SLOT_COMPONENT_TYPEOF_SYMBOL,
```

Comment:
What does `$$typeof`? I removed it together with `SLOT_COMPONENT_TYPEOF_SYMBOL` and everything still works 🤔

Reply:
It is a type that is defined by `React.ExoticComponent`, every single exotic component has that property to help react to identify what component type is that. We can remove it for sure, I just kept it there to maintain implementation similar to what react does.

Reply:
Let's remove it as it does nothing for this implementation.